modulus/datapipes/climate/climate.py

```diff
@@ -0,0 +1,829 @@
+
+
+class ClimateDatapipe(Datapipe):
+    """
+    A Climate DALI data pipeline. This pipeline loads data from
+    HDF5/NetCDF4 files, which can include latitude, longitude, cosine of the
+    solar zenith angle, geopotential, and land-sea mask if specified. Additionally,
+    it normalizes
+    the data if a statistics file is provided. The pipeline returns a dictionary
+    with the following structure, where {name} indicates the name of the data
+    source provided:
+
+    - `state_seq-{name}`: Tensors of shape
+      (batch_size, num_steps, num_channels, height, width).
+      This sequence is drawn from the data file and normalized if a
+      statistics file is provided.
+    - `timestamps-{name}`: Tensors of shape (batch_size, num_steps), containing
+      timestamps for each timestep in the sequence.
+    - `land_sea_mask`: Tensor of shape (batch_size, 1, height, width),
+      containing the land sea mask if a path to a land sea mask file is
+      provided.
+    - `geopotential`: Tensor of shape (batch_size, 1, height, width), containing
+      geopotential if a path to a geopotential file is provided.
+    - `latlon`: Tensor of shape (batch_size, 2, height, width), containing
+      latitude and longitude meshgrid if specified.
+    - `cos_latlon`: Tensor of shape (batch_size, 3, height, width), containing
+      `[cos(lat), sin(lon), cos(lon)]` if specified. This is required by many
+      neural climate models.
+    - `cos_zenith-{name}`: Tensors of shape (batch_size, num_steps, 1, height, width),
+      containing the cosine of the solar zenith angle if specified.
+
+    To use this data pipeline, your data directory must be structured as
+    follows:
+    ```
+    data_dir
+    ├── 1980.h5
+    ├── 1981.h5
+    ├── 1982.h5
+    ├── ...
+    └── 2020.h5
+    ```
+
+    The files are assumed have no metadata, such as timestamps.
+    Because of this, it's important to specify the `dt` parameter and the
+    `start_year` parameter so that the pipeline can compute the correct
+    timestamps for each timestep. These timestamps are then used to compute the
+    cosine of the solar zenith angle, if specified.
+
+    Parameters
+    ----------
+    sources: Iterable[ClimateDataSpec]
+        A list of data specifications defining the sources for the climate variables
+    batch_size : int, optional
+        Batch size, by default 1
+    dt : float, optional
+        Time in hours between each timestep in the dataset, by default 6 hr
+    start_year : int, optional
+        Start year of dataset, by default 1980
+    lsm_filename : str, optional
+        Path to land sea mask file, by default None
+    lsm_variable : str, optional
+        Variable to load from land sea mask file, by default "LSM"
+    geopotential_filename : str, optional
+        Path to geopotential file, by default None
+    geopotential_variable : str, optional
+        Variable to load from geopotential file, by default "Z"
```

Comment:
Could a dictionary-based strategy be used here instead?

Reply:
Yes, these can be in a dictionary. Would it work to have the auxiliary variables configured with a dict mapping variable names to callables that provide them?

Reply:
A dict-based interface to adding auxiliary variables is now implemented in e3a916f6a3b9bb4984f13b67fc53a41fe884624b.